// mongoAdapter/model.js
/**
 * Created with JetBrains WebStorm.
 * User: lee
 * Date: 6/20/13
 * Time: 1:58 PM
 */

(function (mongoose) {

	// Database
	var connection = mongoose.connect('mongodb://localhost/graph-o-matic');

	var Schema = mongoose.Schema;
	var ObjectId = mongoose.Schema.Types.ObjectId;
	var itemData = new Schema({
		name: { type: String,  required: true },
		value: { type: String, required: true },
		type: { type: String }
	});
    var ItemDataSchema = new Schema({
        item: { type: ObjectId, ref: 'Item', required: true},
        data: [itemData]
    });
    var schemas = [];
    schemas.push(ItemDataSchema);

	var originSchema = new Schema({
		context: String,
		area: String,
		vendor: String,
		version: String
	});
    var optionsSchema = new Schema({
        itemLimit: { type:Number, required:false},
        filters:{ type:Object, required:false}
    });

	var CategorySchema = new Schema({
		name: {type:String},
		parentName: {type: String}
		,_origin :[originSchema]
	});
    CategorySchema.index({"name":1, "_origin.context": 1, "_origin.area": 1}, {unique: true});

	schemas.push(CategorySchema);

	var constraintSchema = new Schema({
		To: [
			{type: String  }// itemType
		],
		From: [
			{ type: String  }// itemType
		],
		ToExpression: String,
		FromExpression: String,
		FromToExpression: String
	});

	var RelationshipTypeSchema = new Schema({
		name: {type: String, required: true},
		composite: Boolean,
		bidirectional: Boolean,
		reciprocalRelationship: {type: String },
		parentName: {type: String},
		category: {type: String },
		temporal: Boolean,
		constraints: {
			To: [
				{ type: String }
			],
			From: [
                { type: String }
            ],
			ToExpression: String,
			FromExpression: String,
			FromToExpression: String
		},
		_origin : [originSchema]
	});
    RelationshipTypeSchema.index({"name":1, "origin.context": 1, "origin.area": 1}, {unique: true});

	schemas.push(RelationshipTypeSchema);

	var RelationshipSchema = new Schema({
        typeName: {type: String, required: true},
		type: { type: ObjectId, ref: 'RelationshipType', required: false},
		data: {
			name: { type: String, unique: true, required: true },
			value: { type: String, required: true },
			type: { type: String }
		},
		active: { type: Boolean, required: true },
		fromItem: { type: String, required: true },
		toItem: { type: String, required: true },
		durations: {
			start: Date,
			end: Date
		},
		notes: {
			date: {type: String, required: true},
			note: String
		}
	});
	RelationshipSchema.index({"typeName":1}, {unique: true});

	schemas.push(RelationshipSchema);

	var propertiesSchema = new Schema(
		{
			name: {"required": true, "type":String},
			simpleProperty: {"required": true, "type":"boolean", "default":true},
			isArray: { "type":"boolean", "default":true},
			required: Boolean,

			// THIS
			type: { type:String, enum: ["number", "text", "boolean", "date",
                "email", "currency", "map", "url",
                "atLeastOneOf","oneOf", "oneOrMoreOf" ]},

			//OR THIS
			itemType:{type: String},//ItemType

			description: { type: String, required: false },
            derived: Boolean
		});

	var ItemTypeSchema = new Schema({
		name: { type: String, required: true },
		title: { type: String, required: false },
		description: { type: String, required: false },
		parentName: {type: String},
		category: {type: String },
		properties: [propertiesSchema],
		defaults: [
			{
				name: { type: String, required: true },
				value: { type: String, required: true },
				fixed: Boolean
			}
		],
		allowExtraProperties: Boolean,
		_origin : [originSchema]
	});

	ItemTypeSchema.index({"name":1, "origin.context": 1, "origin.area": 1}, {unique: true});
	ItemTypeSchema.set('versionKey', false);
	schemas.push(ItemTypeSchema);

	var ItemSchema = new Schema({
		typeName: { type:String, required: true },// full name: context.area.name
		type: { type: ObjectId, ref: 'ItemType', required: false },
		title: { type: String, required: true },
		description: { type: String },
		relatedImages: [
			{ type: String }
		],
		relationships: [
			{ type: ObjectId, ref: 'Relationship' }
		], /// ids only
        //data: { type: ObjectId, ref: 'ItemData', required: false },
        data: [{
            name: { type: String, required: true },
            value: { type: String, required: true },
            type: { type: String }
        }]
		//_origin : [originSchema]
	});
	schemas.push(ItemSchema);

	var ViewItemSchema = new Schema({
		item: { type: ObjectId, ref: 'Item', required: true },
		position: {
			x: Number,
			y: Number
		},
        state: String,
		selected: Boolean
	});

	var ContextSchema = new Schema({
		name: {type: String, required: true},
		area: {type: String, required: true},
		types: [String]
	});
	ContextSchema.index({name: 1, area: 1}, {unique: true});
	schemas.push(ContextSchema);

    var ViewTypeSchema = new Schema({
        name: String,
        itemTemplateURL: String,
        relationshipTemplateURL: String,
        cssFiles: [String],
        title: String,
        _origin : [originSchema]
    });
    schemas.push(ViewTypeSchema);


    ////// Summary of a view
     var ViewSummarySchema = new Schema({
        name: {type: String, required: true},
        viewId: {type: String, required: false},
        typeName: {type: String, required: true},
        title: {type: String, required: false},
        itemCount: {type: Number, required:true}
    });
    schemas.push(ViewSummarySchema);

    ///// Data related to view
    var ViewDataSchema = new Schema({
        items: [ViewItemSchema]
    });
    schemas.push(ViewDataSchema);

    ///// The View Object
    /// always has summary, may not have data
    var ViewSchema = new Schema({
        //summary: {ref: 'ViewSummary', type: ObjectId, required: true},
        summary: [ ViewSummarySchema],
        data: {ref: 'ViewData', type: ObjectId, required: false},
        viewOptions:[ optionsSchema ]
    });
    schemas.push(ViewSchema);

    /// MODELS
	this.Category = connection.model('Category', CategorySchema);
	this.ItemType = connection.model('ItemType', ItemTypeSchema);
	this.Item = connection.model('Item', ItemSchema);
	this.ItemData = connection.model('ItemData', ItemDataSchema);
	this.Relationship = connection.model('Relationship', RelationshipSchema);
	this.RelationshipType = connection.model('RelationshipType', RelationshipTypeSchema);
	this.ViewType = connection.model('ViewType', ViewTypeSchema);
	this.ViewSummary = connection.model('ViewSummary', ViewSummarySchema);
	this.ViewData = connection.model('ViewData', ViewDataSchema);
	this.View = connection.model('View', ViewSchema);
	this.ViewItem = connection.model('ViewItem', ViewItemSchema);
	this.Contexts = connection.model('Contexts', ContextSchema);

	/////////// EXPORTS ////////////
	/////////// EXPORTS ////////////
	exports.Category = this.Category;
	exports.ItemType = this.ItemType;
	exports.ItemData = this.ItemData;
	exports.Item = this.Item;
	exports.Relationship = this.Relationship;
	exports.RelationshipType = this.RelationshipType;
	exports.ViewType = this.ViewType;
	exports.ViewSummary = this.ViewSummary;
	exports.ViewData = this.ViewData;
	exports.View = this.View;
	exports.ViewItem = this.ViewItem;
	exports.Contexts = this.Contexts;
	exports.ObjectId = mongoose.Types.ObjectId;

})(require("mongoose"));
